### PR TITLE
Fix ReDoS-bug in string_helper.php

### PR DESCRIPTION
**Pull request by Calico90**
Fix for ReDoS (Regular Expression Denial of Service) / Code Injection Risk as described here: 
https://github.com/bcit-ci/CodeIgniter/issues/4023

### CODE REVIEW

File: system/helpers/string_helper.php [hunk 1]
```diff
@@ -242,7 +242,7 @@ function random_string($type = 'alnum', $len = 8)
 	 */
 	function increment_string($str, $separator = '_', $first = 1)
 	{
-		preg_match('/(.+)'.$separator.'([0-9]+)$/', $str, $match);
+		preg_match('/(.+)' . preg_quote($separator) . '([0-9]+)$/', $str, $match);
```
Comment (narfbg) [inline system/helpers/string_helper.php]: Should be `preg_quote($separator, '/')` instead, to account for the specific delimiter. Also, please remove the spaces around concatenation.

**Comment by narfbg** [comment]
This would also need a changelog entry, but please read my comment on #4023 before marking it as a security bugfix. :)

**Comment by Calico90** [comment]
I'll prepare a new PR. Good point about the `preg_quote` delimiter.

**Comment by narfbg** [comment]
No need, just push the changes to the same branch. Almost every PR goes through some changes, we'd have 40000 of them if they were resubmitted anew. :)

**Comment by Calico90** [comment]
Check, just a branch conflict now. Can someone (with write access) fix it? Thanks

**Comment by narfbg** [comment]
We can't fix conflicts, only you can.

This conflict appeared because you've based you patch on a very, very old version of our develop branch (what you have on line 127, we have at 208 - that's a lot of changes).  
In order to resolve the conflict, you'd neet to do a hard reset back to the last commit that you've gotten from our repository, then pull all of our changes since and commit your changes again.

... or delete your fork and re-fork our repo, in which case you probably would need to open a new pull request. :)

Nevertheless, you're now using `preg_quote()` not just inside the regular expression, but also in the return value, which shouln't happen. And as per my commens in #4023, I still insist against labeling this as a vulnerability (which in fact you're doing twice).

If you don't mind, I'd commit the fix & changelog message myself. Unless you delete your fork, I'll cherry-pick your first commit so you'd be properly attributed in the git log.

**Comment by Calico90** [comment]
@narfbg I don't mind if you commit the fix. My bad about the return, I misread the part before else `:` as if it was the condition of the if statement, not the returning value. But it's obvious a shorthand if statement. Thanks! Won't label it as a security vulnerability.

**Comment by narfbg** [comment]
Done, thanks. :)

**Comment by Calico90** [comment]
@narfbg Good call! This will clear things up! :)

**Pull request closed without merging**